### PR TITLE
WIP: Add probe for HTTP requests with Seahorse

### PR DESCRIPTION
**Pull request by franzliedke**
Looking for early input. It's hard to find out how to hook into Seahorse correctly - I did it by reverse-engineering; maybe I should have searched for documentation more thoroughly.

### CODE REVIEW

**Comment by franzliedke** [comment]
Which tests would you like to see?

File: lib/mnemosyne/middleware/seahorse.rb [hunk 1]
```diff
@@ -0,0 +1,41 @@
+# frozen_string_literal: true
+
+require 'faraday'
```
Comment (jgraichen) [inline lib/mnemosyne/middleware/seahorse.rb]: Left over?

File: spec/mnemosyne/probes/seahorse/base_spec.rb [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+    it 'creates a span when tracing' do
+      trace = with_trace { request }
+
+      expect(trace.span.size).to eq 1
+
+      span = trace.span.first
+
+      expect(span.name).to eq 'external.http.seahorse'
+      expect(span.meta[:url]).to eq 'http://s3.example.com'
```
Comment (jgraichen) [inline spec/mnemosyne/probes/seahorse/base_spec.rb]: Shouldn't the URL include bucket and key from the operation definition above?

File: lib/mnemosyne/probes/seahorse/base.rb [hunk 3]
```diff
@@ -0,0 +1,19 @@
+# frozen_string_literal: true
+
+module Mnemosyne
+  module Probes
+    module Seahorse
+      module Base
+        class Probe < ::Mnemosyne::Probe
+          def setup
+            require 'mnemosyne/middleware/seahorse'
```
Comment (jgraichen) [inline lib/mnemosyne/probes/seahorse/base.rb]: Hmm, do we want to name it "handler" an place inside the probe directory?

I'm not a fan of the global middleware directory anymore...

**Comment by jgraichen** [comment]
Please make the version configurable

File: lib/mnemosyne/middleware/seahorse.rb (hunk 1 above)
Comment (franzliedke) [reply lib/mnemosyne/middleware/seahorse.rb:3]: Yep, removed.

File: lib/mnemosyne/probes/seahorse/base.rb (hunk 3 above)
Comment (franzliedke) [reply lib/mnemosyne/probes/seahorse/base.rb:9]: Yeah, makes sense. I named it "Plugin", as that's what we need to hook into Seahorse.

**Comment by franzliedke** [comment]
I took care of almost all but one comment and added a line to the build configuration.

I am now working on adding more test cases, and actually testing this on my local system.

File: spec/mnemosyne/probes/seahorse/base_spec.rb (hunk 2 above)
Comment (franzliedke) [reply spec/mnemosyne/probes/seahorse/base_spec.rb:40]: One could think so. :thinking: 

To be honest, I don't know enough about how this is supposed to work. I'll have another look.

Comment (Dome-GER) [reply spec/mnemosyne/probes/seahorse/base_spec.rb:40]: Would be nice to have it I suppose. But don't know much about the internals as well.

**Comment by franzliedke** [comment]
Update: @Dome-GER and I found a way to hook into the AWS SDK in such a way that even requests with their S3 API client are properly instrumented.

However, the integration is quite brittle in that it depends on the load order of `aws-sdk-core`, `aws-sdk-s3` and `mnemosyne` itself.

We tabled this for now and instead explicitly added the instrumentation to our internal S3 wrapper, but we could continue integration here at a later date, e.g. by going the monkey-patching route instead.

**Comment by jgraichen** [comment]
@franzliedke I'd fine with adding e.g. a seahorse middleware that needs to be installed manually but keeping and testing it in the client gem.

**Comment by franzliedke** [comment]
I pushed the current version that tries to auto-inject. The test will succeed when the example with tracing runs before the untraced one, but not vice versa. :speak_no_evil:

**Comment by martinstiago** [comment]
@franzliedke, could you provide a list of missing tasks for this MR to remove the WIP status from it? It would help others to jump into the issue.

**Comment by franzliedke** [comment]
Hi @martinstiago, thanks for looking into it!

Basically, what's left is this:

> The test will succeed when the example with tracing runs before the untraced one, but not vice versa.

That's currently the thing that's wrong about this PR, and needs to be fixed. Either that, or instead don't do auto-injection and require consumers of this library to manually enable Seahorse tracing. @jgraichen said he would be fine with that. Would need to be documented in the README as well.

**Comment by jgraichen** [comment]
You might need to change the PR target to the `main` branch as of now.

**Pull request closed without merging**